lib/galaxy/managers/workflows.py

```diff
@@ -403,23 +408,29 @@ def update_workflow_from_raw_description(self, trans, stored_workflow, raw_workf
 
         # Put parameters in workflow mode
         trans.workflow_building_mode = workflow_building_modes.ENABLED
-
+        dry_run = workflow_update_options.dry_run
         workflow, missing_tool_tups = self._workflow_from_raw_description(
             trans,
             raw_workflow_description,
             workflow_update_options,
             name=stored_workflow.name,
+            dry_run=dry_run,
         )
 
-        if missing_tool_tups:
+        if missing_tool_tups and not workflow_update_options.allow_missing_tools:
             errors = []
             for missing_tool_tup in missing_tool_tups:
                 errors.append("Step %i: Requires tool '%s'." % (int(missing_tool_tup[3]) + 1, missing_tool_tup[0]))
             raise MissingToolsException(workflow, errors)
 
         # Connect up
-        workflow.stored_workflow = stored_workflow
-        stored_workflow.latest_workflow = workflow
+        if not dry_run:
+            workflow.stored_workflow = stored_workflow
+            stored_workflow.latest_workflow = workflow
+        else:
+            stored_workflow = model.StoredWorkflow()  # detached
```

Comment:
I did extensive testing - I found all the places for these actions at least and there were some tricky ones. But in the dry run tests, I flushed the session, performed the actions, flush, and then made sure none of the workflow objects in the database were touched. If we do dry run tests for each new refactoring I think this is fine.

I really think the biggest problem is just the lib/galaxy/workflow/modules.py must injects a bunch of random state onto workflow steps that isn't defined at the model layer. If we had a clean module layer and stateless modules I don't think this would be so bad. Pydanic models that mirror the sqlite ones would be fine also - we probably should have a best practice for doing that.

Reply:
xref https://github.com/galaxyproject/galaxy/pull/11032/files#diff-346d4ccd160634059af0bbdd140037c890f4974b57cf10ee1f88eff0d1576f2fR727